packages/circuits/test/bigComparators.test.ts

```diff
@@ -141,16 +141,16 @@ describe('Compare large numbers', function () {
 
     it('should compare upper bits', async () => {
         for (let x = 0; x < 20; x++) {
-            const n1 = random()
-            const n2 = random()
+            const n1 = random() % BigInt(2) ** BigInt(253)
+            const n2 = random() % BigInt(2) ** BigInt(253)
             const { isValid, publicSignals } = await genProofAndVerify(
                 'upperComparators' as any,
                 {
                     in: [n1.toString(), n2.toString()],
                 }
             )
-            const n1Upper = n1 >> BigInt(190)
-            const n2Upper = n2 >> BigInt(190)
+            const n1Upper = n1 >> BigInt(189)
+            const n2Upper = n2 >> BigInt(189)
             const result = n1Upper < n2Upper ? '1' : '0'
```

Comment:
I think the number is confusing
I prefer to change the circuit instead of changing the bits

or just keep it `Num2Bits_strict`

Reply:
I don't understand what you mean, change the circuit where?

Reply:
If we do like

```
signal shifted_value <-- in >> (253-n);
// check the left shift using multiplication
in === shifted_value * 2**(253-n);
// constrain the range of shifted_value to prevent overflow
Num2Bits(n)([shifted_value]);
```

I thought this wouldn't work, but looking at it again maybe it does? We can't use the full 254 bits because the field is actually ~253.7 bits, so small values could be overflowed even with the last range check. This should be ~5x cheaper than doing `Num2Bits(253)`

Reply:
Actually this doesn't work because it fails to account for the lower bits in the left shift check.

Based on [this](https://crypto.stackexchange.com/questions/98415/how-to-generate-constraint-on-right-shift-bitwise-operator-in-circom) answer we could shift then range check the lower bits, bit this only saves ~400 constraints.

It would be most efficient to re-arrange the layout so that the nonce is in the lower bits. Then we'd only need to check 48 bits instead of 205. This would cut the constraints to `24,462`.